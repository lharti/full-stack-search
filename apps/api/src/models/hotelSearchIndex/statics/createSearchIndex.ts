import { hotelsSemanticTokens } from '@/data/hotelsSemanticTokens'
import { Hotel } from '@/models/hotel'
import { HotelSearchIndex } from '@/models/hotelSearchIndex'
import { ReturnModelType } from '@typegoose/typegoose'
import { trigram } from 'n-gram'

export async function createSearchIndex(
    this: ReturnModelType<typeof HotelSearchIndex>,
    hotel: Hotel,
) {
    const { prefixTokens, fragmentsTokens, semanticTokens } =
        createSearchTokens(hotel.hotelName, hotel.country, hotel.city)

    return this.create({
        hotelId: hotel._id,
        hotelName: hotel.hotelName,

        prefixTokens,
        fragmentsTokens,
        semanticTokens,
    })
}

// Internal helpers start
const createSearchTokens = (
    hotelName: string,
    country: string,
    city: string,
): {
    prefixTokens: string[]
    fragmentsTokens: string[]
    semanticTokens: string[]
} => {
    const text = hotelName.trim().replace(/\s+/g, ' ').toLowerCase()

    const prefixTokens = createPrefixTokens(text)

    const fragmentsTokens = createFragmentsTokens(text)

    const semanticTokens = createSemanticTokens(hotelName, [
        country.toLowerCase(),
        city.toLowerCase(),
    ])

    return {
        prefixTokens,
        fragmentsTokens,
        semanticTokens,
    }
}

const createPrefixTokens = (text: string): string[] => {
    const characters = text.split('')

    return characters.map((_, index) => text.slice(0, ++index))
}

const createFragmentsTokens = (text: string): string[] => {
    const words = text.split(' ')

    return words
        .map(word => trigram(word))
        .flat()
        .concat(words)
}

const createSemanticTokens = (
    hotelName: string,
    initialTokens: string[],
): string[] => {
    const preComputedTokens =
        hotelsSemanticTokens.find(hotel => hotel.hotelName === hotelName)
            ?.tokens || []

    return Array.from(new Set([...initialTokens, ...preComputedTokens]))
}
// Internal helpers end
