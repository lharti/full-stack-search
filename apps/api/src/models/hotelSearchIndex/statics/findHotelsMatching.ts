import { HotelSearchIndex } from '@/models/hotelSearchIndex/hotelSearchIndex.model'
import { ReturnModelType } from '@typegoose/typegoose'

export async function findHotelsMatching(
    this: ReturnModelType<typeof HotelSearchIndex>,
    searchQuery: string,
) {
    const matchStage = {
        $match: {
            $or: [
                { prefixTokens: searchQuery },
                { fragmentsTokens: searchQuery },
                { semanticTokens: searchQuery },
            ],
        },
    }

    const createMatchScoreStage = {
        $addFields: {
            matchScore: {
                $switch: {
                    branches: [
                        {
                            case: { $in: [searchQuery, '$prefixTokens'] },
                            then: 3,
                        },
                        {
                            case: {
                                $in: [searchQuery, '$fragmentsTokens'],
                            },
                            then: 2,
                        },
                        {
                            case: { $in: [searchQuery, '$semanticTokens'] },
                            then: 1,
                        },
                    ],

                    default: 0,
                },
            },
        },
    }

    const sortByMatchScoreStage = {
        $sort: {
            matchScore: -1,
            hotelName: 1,
        },
    } as const

    const projectionStage = {
        $project: {
            _id: 0,
            hotelId: 1,
            hotelName: 1,
            matchScore: 1,
        },
    }

    return await this.aggregate([
        matchStage,
        createMatchScoreStage,
        sortByMatchScoreStage,

        projectionStage,
    ])
}
