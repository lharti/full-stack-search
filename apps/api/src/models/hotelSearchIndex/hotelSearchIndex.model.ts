import { getModelForClass, prop, PropType } from '@typegoose/typegoose'
import { Types } from 'mongoose'
import { createSearchIndex } from './statics/createSearchIndex'
import { findHotelsMatching } from './statics/findHotelsMatching'

export class HotelSearchIndex {
    @prop({
        type: String,
        required: true,
    })
    public hotelName!: string

    @prop({
        type: Types.ObjectId,
        required: true,
    })
    public hotelId!: Types.ObjectId

    @prop(
        {
            type: [String],
            required: false,

            index: true,
        },

        PropType.ARRAY,
    )
    public prefixTokens?: string[]

    @prop(
        {
            type: [String],
            required: false,

            index: true,
        },

        PropType.ARRAY,
    )
    public fragmentsTokens?: string[]

    @prop(
        {
            type: [String],
            required: false,

            index: true,
        },

        PropType.ARRAY,
    )
    public semanticTokens: string[] | undefined

    public static createSearchIndex = createSearchIndex
    public static findHotelsMatching = findHotelsMatching
}

export const HotelSearchIndexModel = getModelForClass(HotelSearchIndex)
