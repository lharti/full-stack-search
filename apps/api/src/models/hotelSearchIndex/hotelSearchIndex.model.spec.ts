import { HotelSearchIndexModel } from './hotelSearchIndex.model'
import { createSearchIndex } from './statics/createSearchIndex'
import { findHotelsMatching } from './statics/findHotelsMatching'

describe('HotelSearchIndexModel', () => {
    it('should have the right collection name', () => {
        expect.assertions(1)

        expect(HotelSearchIndexModel.collection.name).toBe('hotelsearchindexes')
    })

    it('should have the right static methods', () => {
        expect.assertions(2)

        expect(HotelSearchIndexModel.findHotelsMatching).toBe(
            findHotelsMatching,
        )

        expect(HotelSearchIndexModel.createSearchIndex).toBe(createSearchIndex)
    })

    it('should have the right schema object', () => {
        expect.assertions(1)

        expect(HotelSearchIndexModel.schema.obj).toMatchInlineSnapshot(`
                            {
                              "fragmentsTokens": {
                                "index": true,
                                "required": false,
                                "type": [
                                  {
                                    "type": [Function],
                                  },
                                ],
                              },
                              "hotelId": {
                                "required": true,
                                "type": [Function],
                              },
                              "hotelName": {
                                "required": true,
                                "type": [Function],
                              },
                              "prefixTokens": {
                                "index": true,
                                "required": false,
                                "type": [
                                  {
                                    "type": [Function],
                                  },
                                ],
                              },
                              "semanticTokens": {
                                "index": true,
                                "required": false,
                                "type": [
                                  {
                                    "type": [Function],
                                  },
                                ],
                              },
                            }
                    `)
    })

    it('should have the right schema options', () => {
        expect.assertions(1)

        // @ts-expect-error: options exists but not defined in types by mongoose
        expect(HotelSearchIndexModel.schema.options).toMatchInlineSnapshot(`
                            {
                              "_id": true,
                              "autoIndex": null,
                              "bufferCommands": true,
                              "capped": false,
                              "discriminatorKey": "__t",
                              "id": true,
                              "minimize": true,
                              "optimisticConcurrency": false,
                              "pluralization": true,
                              "read": null,
                              "shardKey": null,
                              "strict": true,
                              "strictQuery": false,
                              "typeKey": "type",
                              "validateBeforeSave": true,
                              "validateModifiedOnly": false,
                              "versionKey": "__v",
                            }
                    `)
    })

    it('should have the right schema', () => {
        expect.assertions(1)

        expect(HotelSearchIndexModel.schema).toMatchSnapshot()
    })
})
