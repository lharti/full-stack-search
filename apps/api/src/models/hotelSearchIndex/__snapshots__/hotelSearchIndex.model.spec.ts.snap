// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HotelSearchIndexModel should have the right schema 1`] = `
Schema {
  "$globalPluginsApplied": true,
  "$id": 1,
  "_indexes": [],
  "_searchIndexes": [],
  "_userProvidedOptions": {},
  "aliases": {},
  "callQueue": [],
  "childSchemas": [],
  "inherits": {},
  "mapPaths": [],
  "methodOptions": {
    "typegooseName": undefined,
  },
  "methods": {
    "typegooseName": [Function],
  },
  "nested": {},
  "obj": {
    "fragmentsTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
    "hotelId": {
      "required": true,
      "type": [Function],
    },
    "hotelName": {
      "required": true,
      "type": [Function],
    },
    "prefixTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
    "semanticTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
  },
  "options": {
    "_id": true,
    "autoIndex": null,
    "bufferCommands": true,
    "capped": false,
    "discriminatorKey": "__t",
    "id": true,
    "minimize": true,
    "optimisticConcurrency": false,
    "pluralization": true,
    "read": null,
    "shardKey": null,
    "strict": true,
    "strictQuery": false,
    "typeKey": "type",
    "validateBeforeSave": true,
    "validateModifiedOnly": false,
    "versionKey": "__v",
  },
  "paths": {
    "__v": SchemaNumber {
      "_index": null,
      "_presplitPath": [
        "__v",
      ],
      "getters": [],
      "instance": "Number",
      "options": SchemaNumberOptions {
        "type": [Function],
      },
      "path": "__v",
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "_id": SchemaObjectId {
      "_index": null,
      "_presplitPath": [
        "_id",
      ],
      "defaultValue": [Function],
      "getters": [],
      "instance": "ObjectId",
      "options": SchemaObjectIdOptions {
        "auto": true,
        "type": "ObjectId",
      },
      "path": "_id",
      "setters": [
        [Function],
      ],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "fragmentsTokens": SchemaArray {
      "$embeddedSchemaType": SchemaString {
        "_arrayParentPath": "fragmentsTokens",
        "_arrayPath": "fragmentsTokens.$",
        "_index": null,
        "_presplitPath": [
          "fragmentsTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "fragmentsTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "$isMongooseArray": true,
      "_index": true,
      "_presplitPath": [
        "fragmentsTokens",
      ],
      "caster": SchemaString {
        "_arrayParentPath": "fragmentsTokens",
        "_arrayPath": "fragmentsTokens.$",
        "_index": null,
        "_presplitPath": [
          "fragmentsTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "fragmentsTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "casterConstructor": [Function],
      "defaultValue": [Function],
      "getters": [],
      "instance": "Array",
      "isRequired": false,
      "options": SchemaArrayOptions {
        "index": true,
        "required": false,
        "type": [
          {
            "type": [Function],
          },
        ],
      },
      "path": "fragmentsTokens",
      "schemaOptions": {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "optimisticConcurrency": false,
        "pluralization": true,
        "read": null,
        "shardKey": null,
        "strict": true,
        "strictQuery": false,
        "typeKey": "type",
        "validateBeforeSave": true,
        "validateModifiedOnly": false,
        "versionKey": "__v",
      },
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "hotelId": SchemaObjectId {
      "_index": null,
      "_presplitPath": [
        "hotelId",
      ],
      "getters": [],
      "instance": "ObjectId",
      "isRequired": true,
      "options": SchemaObjectIdOptions {
        "required": true,
        "type": [Function],
      },
      "originalRequiredValue": true,
      "path": "hotelId",
      "requiredValidator": [Function],
      "setters": [],
      "validators": [
        {
          "message": "Path \`{PATH}\` is required.",
          "type": "required",
          "validator": [Function],
        },
      ],
      Symbol(mongoose#schemaType): true,
    },
    "hotelName": SchemaString {
      "_index": null,
      "_presplitPath": [
        "hotelName",
      ],
      "enumValues": [],
      "getters": [],
      "instance": "String",
      "isRequired": true,
      "options": SchemaStringOptions {
        "required": true,
        "type": [Function],
      },
      "originalRequiredValue": true,
      "path": "hotelName",
      "regExp": null,
      "requiredValidator": [Function],
      "setters": [],
      "validators": [
        {
          "message": "Path \`{PATH}\` is required.",
          "type": "required",
          "validator": [Function],
        },
      ],
      Symbol(mongoose#schemaType): true,
    },
    "prefixTokens": SchemaArray {
      "$embeddedSchemaType": SchemaString {
        "_arrayParentPath": "prefixTokens",
        "_arrayPath": "prefixTokens.$",
        "_index": null,
        "_presplitPath": [
          "prefixTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "prefixTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "$isMongooseArray": true,
      "_index": true,
      "_presplitPath": [
        "prefixTokens",
      ],
      "caster": SchemaString {
        "_arrayParentPath": "prefixTokens",
        "_arrayPath": "prefixTokens.$",
        "_index": null,
        "_presplitPath": [
          "prefixTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "prefixTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "casterConstructor": [Function],
      "defaultValue": [Function],
      "getters": [],
      "instance": "Array",
      "isRequired": false,
      "options": SchemaArrayOptions {
        "index": true,
        "required": false,
        "type": [
          {
            "type": [Function],
          },
        ],
      },
      "path": "prefixTokens",
      "schemaOptions": {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "optimisticConcurrency": false,
        "pluralization": true,
        "read": null,
        "shardKey": null,
        "strict": true,
        "strictQuery": false,
        "typeKey": "type",
        "validateBeforeSave": true,
        "validateModifiedOnly": false,
        "versionKey": "__v",
      },
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "semanticTokens": SchemaArray {
      "$embeddedSchemaType": SchemaString {
        "_arrayParentPath": "semanticTokens",
        "_arrayPath": "semanticTokens.$",
        "_index": null,
        "_presplitPath": [
          "semanticTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "semanticTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "$isMongooseArray": true,
      "_index": true,
      "_presplitPath": [
        "semanticTokens",
      ],
      "caster": SchemaString {
        "_arrayParentPath": "semanticTokens",
        "_arrayPath": "semanticTokens.$",
        "_index": null,
        "_presplitPath": [
          "semanticTokens",
        ],
        "enumValues": [],
        "getters": [],
        "instance": "String",
        "options": SchemaStringOptions {},
        "path": "semanticTokens",
        "regExp": null,
        "setters": [],
        "validators": [],
        Symbol(mongoose#schemaType): true,
      },
      "casterConstructor": [Function],
      "defaultValue": [Function],
      "getters": [],
      "instance": "Array",
      "isRequired": false,
      "options": SchemaArrayOptions {
        "index": true,
        "required": false,
        "type": [
          {
            "type": [Function],
          },
        ],
      },
      "path": "semanticTokens",
      "schemaOptions": {
        "_id": true,
        "autoIndex": null,
        "bufferCommands": true,
        "capped": false,
        "discriminatorKey": "__t",
        "id": true,
        "minimize": true,
        "optimisticConcurrency": false,
        "pluralization": true,
        "read": null,
        "shardKey": null,
        "strict": true,
        "strictQuery": false,
        "typeKey": "type",
        "validateBeforeSave": true,
        "validateModifiedOnly": false,
        "versionKey": "__v",
      },
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
  },
  "plugins": [
    {
      "fn": [Function],
      "opts": {
        "deduplicate": true,
      },
    },
    {
      "fn": [Function],
      "opts": {
        "deduplicate": true,
      },
    },
    {
      "fn": [Function],
      "opts": {
        "deduplicate": true,
      },
    },
    {
      "fn": [Function],
      "opts": {
        "deduplicate": true,
      },
    },
    {
      "fn": [Function],
      "opts": {
        "deduplicate": true,
      },
    },
  ],
  "query": {},
  "s": {
    "hooks": Kareem {
      "_posts": Map {
        "save" => [
          {
            "fn": [Function],
          },
          {
            "fn": [Function],
          },
          {
            "fn": [Function],
          },
        ],
        "init" => [
          {
            "fn": [Function],
          },
        ],
      },
      "_pres": Map {
        "save" => [
          {
            "fn": [Function],
            "isAsync": false,
          },
          {
            "fn": [Function],
            "isAsync": false,
          },
          {
            "fn": [Function],
            "isAsync": false,
          },
          {
            "fn": [Function],
            "isAsync": false,
          },
        ],
        "remove" => [
          {
            "fn": [Function],
            "isAsync": false,
          },
        ],
      },
    },
  },
  "singleNestedPaths": {},
  "statics": {
    "createSearchIndex": [Function],
    "findHotelsMatching": [Function],
  },
  "subpaths": {
    "fragmentsTokens.$": SchemaString {
      "_arrayParentPath": "fragmentsTokens",
      "_arrayPath": "fragmentsTokens.$",
      "_index": null,
      "_presplitPath": [
        "fragmentsTokens",
      ],
      "enumValues": [],
      "getters": [],
      "instance": "String",
      "options": SchemaStringOptions {},
      "path": "fragmentsTokens",
      "regExp": null,
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "prefixTokens.$": SchemaString {
      "_arrayParentPath": "prefixTokens",
      "_arrayPath": "prefixTokens.$",
      "_index": null,
      "_presplitPath": [
        "prefixTokens",
      ],
      "enumValues": [],
      "getters": [],
      "instance": "String",
      "options": SchemaStringOptions {},
      "path": "prefixTokens",
      "regExp": null,
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
    "semanticTokens.$": SchemaString {
      "_arrayParentPath": "semanticTokens",
      "_arrayPath": "semanticTokens.$",
      "_index": null,
      "_presplitPath": [
        "semanticTokens",
      ],
      "enumValues": [],
      "getters": [],
      "instance": "String",
      "options": SchemaStringOptions {},
      "path": "semanticTokens",
      "regExp": null,
      "setters": [],
      "validators": [],
      Symbol(mongoose#schemaType): true,
    },
  },
  "tree": {
    "__v": [Function],
    "_id": {
      "auto": true,
      "type": "ObjectId",
    },
    "fragmentsTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
    "hotelId": {
      "required": true,
      "type": [Function],
    },
    "hotelName": {
      "required": true,
      "type": [Function],
    },
    "id": VirtualType {
      "getters": [
        [Function],
      ],
      "options": {},
      "path": "id",
      "setters": [],
    },
    "prefixTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
    "semanticTokens": {
      "index": true,
      "required": false,
      "type": [
        {
          "type": [Function],
        },
      ],
    },
  },
  "virtuals": {
    "id": VirtualType {
      "getters": [
        [Function],
      ],
      "options": {},
      "path": "id",
      "setters": [],
    },
  },
}
`;
